src/librustc/ty/instance.rs

```diff
@@ -290,6 +298,24 @@ impl<'tcx> Instance<'tcx> {
         result
     }
 
+    pub fn resolve_for_fn_ptr(
+        tcx: TyCtxt<'tcx>,
+        param_env: ty::ParamEnv<'tcx>,
+        def_id: DefId,
+        substs: SubstsRef<'tcx>,
+    ) -> Option<Instance<'tcx>> {
+        debug!("resolve(def_id={:?}, substs={:?})", def_id, substs);
+        if tcx.codegen_fn_attrs(def_id).flags.contains(CodegenFnAttrFlags::TRACK_CALLER) {
```

Comment:
I think you should do this after resolving, so you check the final `def_id`, not the initial one.

Reply:
So, instead of recursing, run Instance::resolve on the `def_id`, check its codegen attrs then, and decide whether to shim it?

Reply:
I did this, lmk if it looks right?